### ISSUE TITLE
Postgresql dialect use of exclusive locks on every startup hangs on modestly sized production instances

### ISSUE THREAD

**Issue by kevinmershon** [xtdb/crux#2582]
This code in the dialect implementation is dropping and recreating database indexes on every startup, and exclusively locking the whole `tx_events` table.
https://github.com/xtdb/xtdb/blob/master/modules/jdbc/src/xtdb/jdbc/psql.clj#L36-L43

For a production database with ~100 million transactions, this is excruciatingly slow, and causes major downtime when one of our several replicas reboots.

**Comment by jarohen** [comment]
Hey @kevinmershon 🙂 

It looks like these statements are referring to two different index names - the first is removing a legacy index if it exists; the second creating the new one if it doesn't. Under normal operation this should be a fast no-op after the first time they're run (first introduced in 1.17.0) - are you still seeing this in your Postgres running queries table?

James

**Comment by danmason** [comment]
Hi there :wave: 

Just to add to what James is saying above - I spun up a postgres backed node locally, transacted a thousand documents onto it, and tracked the logging from my local postgress when the node was restarted (ie, when the call to `setup-schema!` should run) - the following log is from all of the queries being ran in `setup-schema!` (the first one being the 'create table if not exists', the latter two being the 'drop index if exists' and 'create index if not exists'):
```
2023-06-22 11:20:11.622 BST [30899] NOTICE:  relation "tx_events" already exists, skipping
2023-06-22 11:20:11.630 BST [30899] NOTICE:  index "tx_events_event_key_idx" does not exist, skipping
2023-06-22 11:20:11.631 BST [30899] NOTICE:  relation "tx_events_event_key_idx_2" already exists, skipping
```

The following is with around half a million documents on it:
```
2023-06-22 11:32:40.390 BST [32307] NOTICE:  relation "tx_events" already exists, skipping
2023-06-22 11:32:40.391 BST [32307] NOTICE:  index "tx_events_event_key_idx" does not exist, skipping
2023-06-22 11:32:40.391 BST [32307] NOTICE:  relation "tx_events_event_key_idx_2" already exists, skipping
```

**Comment by kevinmershon** [comment]
![image](https://github.com/xtdb/xtdb/assets/812577/e07e5556-46a7-429d-a38a-450d1b7028ac)

We are observing the tx_events_event_key_idx_2 create if not exists query run indefinitely (the index does already exist on the table) on every startup.

I have cloned the namespace to our own codebase, commented out these lines, and gave it a unique name for the dialect to get us back online in production. We suffered 4.5 hours of full system downtime yesterday because of this.

**Comment by jarohen** [comment]
@kevinmershon It unfortunately looks like `CREATE INDEX IF NOT EXISTS` still takes out an exclusive lock on the table even if the index exists, which would cause the issue here. Normally I'd imagine this check runs fast enough that there's a big enough write gap in most use-cases for it to slip through.

`CREATE INDEX CONCURRENTLY IF NOT EXISTS` doesn't take out the same exclusive lock if the index doesn't exist - would this work for your use case? If so, we could release a snapshot for you to use.

(`DROP INDEX IF EXISTS` doesn't appear to take out the lock if the index doesn't exist, so this shouldn't be an issue.)

James

**Comment by kevinmershon** [comment]
> @kevinmershon It unfortunately looks like `CREATE INDEX IF NOT EXISTS` still takes out an exclusive lock on the table even if the index exists, which would cause the issue here. Normally I'd imagine this check runs fast enough that there's a big enough write gap in most use-cases for it to slip through.

Yeah, that makes more sense and jibes with what we're seeing. We run 3 replicas with 1 acting as write master and all 3 supporting queries/reads. With the ongoing [Lucene indexing lock-up bug](https://github.com/xtdb/xtdb/issues/1916) we see 2-3 forced reboots a day per replica and it's been getting slower and slower. At least now I know why.

> `CREATE INDEX CONCURRENTLY IF NOT EXISTS` doesn't take out the same exclusive lock if the index doesn't exist - would this work for your use case? If so, we could release a snapshot for you to use.

No need to release a snapshot, I can modify the query on my cloned namespace and let you know if it works.

**Comment by refset** [comment]
>I can modify the query on my cloned namespace and let you know if it works.

Hey @kevinmershon did you have any joy on this on front?

**Comment by kevinmershon** [comment]
> > I can modify the query on my cloned namespace and let you know if it works.
> 
> Hey @kevinmershon did you have any joy on this on front?

Ah sorry, I forgot to do this. We've been running in production with a redeclaration of the `xtdb.jdbc.psql/->dialect` function that comments out that line entirely.

I'm not in a position to risk our production environments to any downtime for awhile, and our testing environment doesn't have sufficient data to cause the hang. It will be some time before I am able to run this experiment.

**Comment by kevinmershon** [comment]
Giving an update that we also had to remove [this lock](https://github.com/xtdb/xtdb/blob/master/modules/jdbc/src/xtdb/jdbc/psql.clj#L43) today as well after we had a spontaneous multi-replica lockup again.

For systems that are continuously writing, have multiple replicas, and very large tx_events tables, exclusive locks are not ever appropriate.

**Comment by jarohen** [comment]
splitting out into #2687 as I suspect this is a separate issue
